Sources/BSON/BSONDecimal128.swift

```diff
@@ -0,0 +1,517 @@
+        if exponentPartRange.location != NSNotFound, let range = Range(exponentPartRange, in: data) {
+            exponent = exponentSign * (Int(data[range]) ?? 0)
+        }
+        if let pointIndex = decimalPart.firstIndex(of: ".") {
+            // move the exponent by the number of digits after the decimal point
+            // so we are looking at an "integer" significand, easier to reason about
+            exponent -= decimalPart.distance(from: pointIndex, to: decimalPart.endIndex) - 1
+        }
+
+        while exponent > Self.exponentMax && digits.count <= Self.maxSignificandDigits {
+            // Exponent is too large, try shifting zeros into the coefficient
+            digits.append(0)
+            exponent -= 1
+        }
+
+        while exponent < Self.exponentMin && !digits.isEmpty {
+            // Exponent is too small, try taking zeros off the coefficient
+            if digits.count == 1 && digits[0] == 0 {
+                exponent = Self.exponentMin
+                break
+            }
+
+            if digits.last == 0 {
+                digits.removeLast()
+                exponent += 1
+                continue
+            }
+
+            if digits.last != 0 {
+                // We don't end in a zero and our exponent is too small
+                throw BSONError.InvalidArgumentError(message: "Underflow Error: Value too small")
+            }
+        }
+
+        guard (Self.exponentMin...Self.exponentMax).contains(exponent) else {
+            throw BSONError.InvalidArgumentError(message: "Rounding Error: Cannot round exponent \(exponent) further")
+        }
+
+        guard digits.count <= Self.maxSignificandDigits else {
+            throw BSONError.InvalidArgumentError(
+                message: "Overflow Error: Value cannot exceed \(Self.maxSignificandDigits) digits"
+            )
+        }
+
+        let significandLoDigits = [UInt8](digits.suffix(Self.maxSignificandDigits / 2)).decimalDigitsToUInt64()
+        let significandHiDigits = [UInt8](digits.dropLast(Self.maxSignificandDigits / 2)).decimalDigitsToUInt64()
+
+        // Multiply by one hundred quadrillion (note the seventeen zeroes)
+        // the product is the significandHiDigits "shifted" up by 17 decimal places
+        // we can then add the significandLoDigits to the product to ensure that we have a correctly formed significand
+        var significand = UInt128.multiply(significandHiDigits, by: Self.decimalShift17Zeroes)
+        significand.lo += significandLoDigits
+
+        if significand.lo < significandLoDigits {
+            // carry over addition to hi side
+            significand.hi += 1
+        }
+
+        let biasedExponent = UInt64(exponent + Self.exponentBias).getBits(49...63)
+
+        self.value = UInt128()
+
+        // The most significant bit of the significand determines the format
+        // Encode combination, exponent, and significand.
+        if significand.hi.getBit(14) == 1 {
```

Comment:
why is the 14th bit of significand.hi the MSB? I thought the significand was the lowest 110 bits so it would be stored in `significand.hi[18...]` plus all 64 bits of `significand.lo` ?

Reply:
Patrick, you're correct that `0b100` is the pattern to look for but its also true that only 1 has to be seen in this position, however I do think it is clearer to check against 0b100

Kaitlin, The significand is indeed in the lower 110 bits when finally encoded which is what the code inside this if block will do, but upon calculation it can take up more space than that, that's where this idea of implicit bits come from, to increase the encoding's capabilities.

Reply:
Why can't the three most significant bits of the significand be `0b111` or `0b101`? Are we supposed to be looking at the bits of the most significant digit? That would make sense with 9 (`0b101`) being the maximum value. Looking at the wikipedia page, it states that the "most significant 4 bits of the significand" are what we should be looking at, but [this page](http://speleotrove.com/decimal/dbspec.html) refers to "the the most significant digit (MSD) of the coefficient (4 bits, which may take only the values 0 through 9)". Checking the bits of the digit doesn't make a lot of sense to me though because the digits themselves aren't stored--the binary representation of the actual number is.

Reply:
This [page](https://en.wikipedia.org/wiki/Binary_integer_decimal) goes into a little more detail, but it still doesn't mention why `0b111` cannot be the leading bits.

Reply:
The digits are what is being stored not the binary representation of them, each four bit sequence can only hold numbers 0-9. You can see this is being done in the encoding here:
https://github.com/mongodb/swift-bson/blob/cf2670520543749d88fbd5990c00414c742d2b05/Sources/BSON/BSONDecimal128.swift#L57-L61
I take the existing value and multiply by 10 (like a decimal shift, 8 becomes 80)
I then add in the next digit (like a decimal or since digits are guaranteed to be 0-9, 80 becomes 80 + value (does not exceed 89))

Reply:
Right, but you're doing numerical addition and storing it in a `UInt64`, which is binary storage. If you were storing the digits themselves, you'd be implementing the densely packed decimal (DPD) encoding. According to the spec (and your implementation), we're doing the Binary Integer Decimal (BID) encoding instead though, which just stores the significand as a regular binary integer.

Reply:
Okay, so I think I've figured out the answers to my questions, and the answers are very unsatisfying. 

The decimal format itself is defined generically in terms of a _k_ which is a multiple of 32. decimal32, decimal64, and decimal128 are just three of the first possible values for _k_. So theoretically, one could parse / generate arbitrary precision decimal\<k\>'s using the same code and just changing the input of this _k_. The result of this is that they all use the same "implicit bit" logic in the combination field. The trick that I was missing was that decimal128 _doesn't actually need the extra implicit bits_. The entirety of decimal128's range can fit normally by just encoding the exponent and significand as-is.

The calculation for that is as follows:
- the maximum significand is 34 digits (i.e. 9999...), and it can be represented using 113 bits. 
- the maximum biased exponent is 12287, which can be represented in 14 bits.
    - the two MSB of the maximum biased exponent are `0b10`. I think it is a mathematical property based on how the maximum exponents are derived from _k_ that they are never `0b11`. See the wikipedia page I linked to in a previous comment.
- The trailing significand field is 110 bits long, meaning 3 bits of the significand must be stored in the combination field (113 - 110 = 3). 
- The combination is 17 bits long though, meaning it is big enough to store all three missing bits of the significand and the entire exponent (14 + 3 = 17)
    - This is the implicit leading `0b0` case

For other formats, sometimes the implicit `0b100` will need to be used in order to fit everything. In those cases, due to some other mathematical property (I assume) the first three bits of the max significand are never `0b110` or `0b101` or `0b111`. 

The decimal128 wikipedia page actually calls out that decimal128 doesn't ever require the `0b100` case but is just following the general pattern:

>For the decimal128 format, all of these significands are out of the valid range (they begin with 2^113 > 1.038×1034), and are thus decoded as zero, but the pattern is same as decimal32 and decimal64.

So, in summary, I think we can actually simplify this code a bit and skip the parts about setting `0b11` and the implicit `0b100`, and just encode the exponent and significand directly into the combination field sequentially.

Reply:
Welp... yea I missed that! but it makes sense 128 bits is massive after all 👏  spectacular research, thank you for your efforts